File: src/main/scala/com/hpe/ossm/scluster/selfMonitor/Collector.scala
package com.hpe.ossm.scluster.selfMonitor

import java.text.SimpleDateFormat

import akka.actor.{Actor, Timers}
import akka.cluster.Cluster
import akka.cluster.pubsub.{DistributedPubSub, DistributedPubSubMediator}
import akka.kafka.scaladsl.Consumer.Control
import akka.stream.{ActorMaterializer, Materializer}
import com.hpe.ossm.scala.lang.util.KafkaUtil
import com.hpe.ossm.scluster.messges.{CmdKPIRefresh, Collect, KPIRecord}
import com.typesafe.config.Config
import org.apache.kafka.clients.consumer.ConsumerRecord
import org.apache.kafka.clients.producer.{KafkaProducer, ProducerRecord}
import org.json.JSONException
import org.slf4j.{Logger, LoggerFactory}

import scala.concurrent.duration._
import scala.concurrent.ExecutionContext

/**
 * root class for all collectors of self monitor
 */
abstract class Collector extends Actor with Timers {
    protected val LOGGER: Logger = LoggerFactory.getLogger(this.getClass)
    //list of names of KPI: uses to filter refresh cmd
    val kpiNames: List[String]

    protected val conf: Config = context.system.settings.config
    protected val cluster: Cluster = Cluster(context.system)
    private val mediator = DistributedPubSub(context.system).mediator
    private implicit val ec: ExecutionContext = context.dispatcher
    private implicit val mat: Materializer = ActorMaterializer.create(context.system)
    protected val host: String = cluster.selfAddress.host.orNull
    private val kafkaActive = conf.getBoolean("kafka.active")
    private val topic = conf.getString("ossm.monitor.topic")
    private val key_cmd = conf.getString("ossm.monitor.keys.cmd")
    private val key_record = conf.getString("ossm.monitor.keys.record")
    //Kafka producer to publish KPI
    private var producer: KafkaProducer[String, java.io.Serializable] = _
    private var consumer: Control = _
    private val myPath = self.path.toSerializationFormatWithAddress(cluster.selfAddress)

    override def preStart(): Unit = {
        println(s"Collector $myPath [$kpiNames] on $host starting")
        //        super.preStart()
        cluster.registerOnMemberUp(() => {
            initCollector()
            LOGGER.debug(s"Collector is UP, Kafka: $kafkaActive")
            if (kafkaActive) {
                producer = KafkaUtil.createProcedure(conf)
                consumer = KafkaUtil.createAkkaConsumer(conf, myPath, topic, (msg: ConsumerRecord[String, String]) => {
                    try {
                        msg.key match {
                            case `key_cmd` => self ! CmdKPIRefresh.fromString(msg.value)
                            case _ =>
                        }
                    } catch {
                        case e: JSONException => println(s"${e.getMessage}\n ${msg.value}")
                    }
                }).run
            } else {
                mediator ! DistributedPubSubMediator.Subscribe(topic, self)
            }
        })
    }

    override def postStop(): Unit = {
        super.postStop()
        if (kafkaActive) {
            producer.close()
            consumer.stop()
        } else {
            cluster.unsubscribe(self)
        }
        println(s"Collector $myPath [$kpiNames] on $host stopped")
    }

    override def receive: Receive = {
        case CmdKPIRefresh(kpiName, host, ts) if kpiNames.contains(kpiName) && ((host == null) || host.equals(this.host)) =>
            //make sure the value of 'host' from FE is same as that in BE
            LOGGER.debug(s"Received CMD to refresh the collector ${new SimpleDateFormat("yyyy-MM-dd HH:mm:ss").format(ts)}")
            publish(refreshKPI(kpiName))
        case Collect =>
            LOGGER.debug(s"Received Collect Message")
            publish(collect)
    }

    protected def publish(records: List[KPIRecord]): Unit = {
        if (kafkaActive)
            records.foreach(r =>
                if (r != null) producer.send(new ProducerRecord[String, java.io.Serializable](topic, key_record, r.toString))
                else LOGGER.warn(s"Receive Null KPI")
            )
        else
            records.foreach(r =>
                if (r != null) mediator ! DistributedPubSubMediator.Publish(topic, r)
                else LOGGER.warn(s"Receive Null KPI")
            )
    }

    protected def setTimer(interval: Int): Unit = {
        publish(collect) //run collector at beginning
        if (interval > 0) timers.startPeriodicTimer("collect", Collect, interval.seconds)
    }


    protected def stopSelf(): Unit = context stop self

    //Functions to be implemented in children

    /**
     * Get configurations and start timer to execute the 'collect' function
     *
     */
    def initCollector(): Unit

    /**
     * function to run and publish the configured KPI once
     *
     * @return
     */
    def collect: List[KPIRecord]

    /**
     * function to collect KPI (name as kpiName) on command
     */

    def refreshKPI(kpiName: String): List[KPIRecord]
}

File: src/main/java/com/hpe/ossm/jcluster/seflMonitor/Collector.java
package com.hpe.ossm.jcluster.seflMonitor;

import akka.actor.AbstractActorWithTimers;
import akka.actor.ActorRef;
import akka.cluster.Cluster;
import akka.cluster.pubsub.DistributedPubSub;
import akka.cluster.pubsub.DistributedPubSubMediator;
import akka.stream.ActorMaterializer;
import akka.stream.Materializer;
import com.hpe.ossm.scala.lang.util.KafkaUtil;
import com.hpe.ossm.scluster.messges.CmdKPIRefresh;
import com.hpe.ossm.scluster.messges.KPIRecord;
import com.typesafe.config.Config;
import org.apache.kafka.clients.consumer.ConsumerRecord;
import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.clients.producer.ProducerRecord;
import org.slf4j.Logger;
import akka.kafka.scaladsl.Consumer.Control;//reuse the lib function, or to create for JAVA

import java.io.Serializable;
import java.time.Duration;
import java.util.List;

import org.json.JSONException;
import org.slf4j.LoggerFactory;

abstract public class Collector extends AbstractActorWithTimers {
    protected Logger LOGGER= LoggerFactory.getLogger(this.getClass());
    protected List<String> kpiNames;

    private final Config conf = context().system().settings().config();
    private final Cluster cluster = Cluster.get(context().system());
    private final ActorRef mediator = DistributedPubSub.get(context().system()).mediator();
    private final Materializer mat = ActorMaterializer.create(context().system());
    protected final String host = cluster.selfAddress().host().get();
    private final Boolean kafkaActive = conf.getBoolean("kafka.active");
    private final String topic = conf.getString("ossm.monitor.topic");
    private final String key_cmd = conf.getString("ossm.monitor.keys.cmd");
    private final String key_record = conf.getString("ossm.monitor.keys.record");

    private KafkaProducer<String, Serializable> producer;
    private Control consumer;
    private final String myPath = self().path().toSerializationFormatWithAddress(cluster.selfAddress());

    abstract public void initCollector();

    abstract public List<KPIRecord> collect();

    abstract public List<KPIRecord> refreshKPI(String kpiName);

    @Override
    public void preStart() throws Exception {
        super.preStart();
        LOGGER.info("Collector {} starting", myPath);
        cluster.registerOnMemberUp(() -> {
            initCollector();
            if (kafkaActive) {
                producer = KafkaUtil.createProcedure(conf);
                consumer = KafkaUtil.createAkkaConsumer(conf, myPath, topic, (ConsumerRecord<String, String> msg) -> {
                    try {
                        if (msg.key().endsWith(key_cmd)) {
                            self().tell(CmdKPIRefresh.fromString(msg.value()), self());
                        }
                    } catch (JSONException e) {
                        LOGGER.error("{} | {}", e.getMessage(), msg.value());
                    }
                    return null;
                }).run(mat);
            } else {
                mediator.tell(new DistributedPubSubMediator.Subscribe(topic, self()), self());
            }
        });
    }

    @Override
    public void postStop() throws Exception {
        super.postStop();
        if (kafkaActive) {
            producer.close();
            consumer.stop();
        } else {
            cluster.unsubscribe(self());
        }
        LOGGER.info("Collector {} stopped", myPath);
    }

    protected void publish(List<KPIRecord> records) {
        if (kafkaActive)
            records.forEach(r -> producer.send(new ProducerRecord<String, java.io.Serializable>(topic, key_record, r.toString())));
        else
            records.forEach(r -> mediator.tell(new DistributedPubSubMediator.Publish(topic, r), self()));
    }

    protected void setTimer(int interval) {
        if (interval > 0) timers().startPeriodicTimer("collect", "Collect", Duration.ofSeconds(interval));
    }

    protected void stopSelf() {
        context().stop(self());
    }

    @Override
    public Receive createReceive() {
        return receiveBuilder()
                .match(CmdKPIRefresh.class, cmd -> {
                    if (kpiNames.contains(cmd.kpiName()) && (host.equals(cmd.host())) || "*".equals(cmd.host())) {
                        publish(refreshKPI(cmd.kpiName()));
                    }
                })
                .matchEquals("Collect", cmd -> {
                    publish(collect());
                })
                .build();
    }
}
